KAFKA-10036 Improve handling and documentation of Suppliers

soarez opened the pull request:
Following up on #8752 which seems to have gone stale.

@mjsax can you continue the review?

soarez commented:
@mjsax can something be done to move this forward?

mjsax commented:
@soarez Is this PR a replacement of #8752 and we should close the other PR?

soarez commented:
@mjsax yes, if we can continue the current work here, then we should close it. I don’t think I’m able to add commits to the original PR.

mjsax commented:
There is a checkstyle error:
```
Task :streams:checkstyleTest
12:00:03 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:1873: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
```

If the other PR is void, can you please close it? Thanks.

soarez commented:
Thanks @mjsax . I've addressed the Checkstyle issue. 

I don't think I'm able to close the original PR. In this Github repository I'm only able to close PRs that I've authored.

@sneakyburro maybe you can close #8752? Or finish the work there and I'll close this one.

soarez commented:
@mjsax what can we do to proceed?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -521,6 +521,10 @@ public synchronized StreamsBuilder addStateStore(final StoreBuilder<?> builder)
      * It is not required to connect a global store to {@link Processor Processors}, {@link Transformer Transformers},
      * or {@link ValueTransformer ValueTransformer}; those have read-only access to all global stores by default.
+     * <p>
+     * The supplier should always generate a new instance each time invoking {@link  ProcessorSupplier#get()}. Creating
```
nit `each time {@link  ProcessorSupplier#get()} is called.` (similar elsewhere -- please fix throughout the whole PR)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -521,6 +521,10 @@ public synchronized StreamsBuilder addStateStore(final StoreBuilder<?> builder)
      * or {@link ValueTransformer ValueTransformer}; those have read-only access to all global stores by default.
+     * <p>
+     * The supplier should always generate a new instance each time invoking {@link  ProcessorSupplier#get()}. Creating
+     * a single Processor object and returning the same object reference in {@link ProcessorSupplier#get()} would be
```
`{@link Processor}`(similar elsewhere -- please fix throughout the whole PR)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -2522,8 +2522,12 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * for each input record, it is recommended to use {@link #flatTransform(TransformerSupplier, String...)
      * flatTransform()}.
+     * The supplier should always generate a new instance each time invoking {@link TransformerSupplier#get()}. Creating
+     * a single Transformer object and returning the same object reference in {@link TransformerSupplier#get()} would be
```
`{@link Transformer}` (also elsewhere)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamUtil.java:
```diff
@@ -0,0 +1,38 @@
+    static void checkTransformerSupplier(final TransformerSupplier<?, ?, ?> supplier) {
+        if (supplier.get() == supplier.get()) {
+            throw new IllegalArgumentException("TransformerSupplier generates single transformer reference. Supplier " +
+                    "pattern violated.");
```
I think we can be more elaborate and add something like `TransformerSupplier#get() must return a new Transformer object each time it is called.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyUtil.java:
```diff
@@ -0,0 +1,39 @@
+    public static void checkProcessorSupplier(final ProcessorSupplier<?, ?> supplier) {
+        if (supplier.get() == supplier.get()) {
+            throw new TopologyException("ProcessorSupplier generates single processor reference. Supplier pattern" +
+                    " violated.");
```
As above.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java:
```diff
@@ -175,6 +177,12 @@ public void testAddProcessorWithNullParents() {
+    @Test(expected = TopologyException.class)
+    public void testAddProcessorWithBadSupplier() {
+        final Processor<Object, Object> processor = new MockProcessor<>();
+        builder.addProcessor("processor", () -> processor, (String) null);
```
We should use `aasertThrows` and verify the error message similar to the `TransformerSupplier` test

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -1865,6 +1865,17 @@ public void shouldNotAllowNullNamedOnTransformWithStoreName() {
     }
 
+    @Test
+    public void shouldNotAllowBadTransformerSupplierOnFlatTransform() {
```
It might be nice to replicate this test for all "siblings" of `flatTransform` (including the "Value" variants). -- Also note that there is `KStream#process()` and `StreamsBuilder#addGlobalStateStore), too, that should also be tested.

soarez replied:
The "Value" variants use a `ValueTransformerSupplier` instead of a `TransformerSupplier`. Does it makes sense to extend the supplier usage verification to `ValueTransformerSupplier` as well?

mjsax replied:
Yes, those suppliers must return a new instance on `get()`, too, so the new check should be done for them as well.

soarez replied:
Ok. I just added equivalent documentation changes, checks and tests for `ValueTransformerSupplier` and `ValueTransformerWithKeySupplier` as well in 567d5e0 .

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java:
```diff
@@ -175,6 +177,12 @@ public void testAddProcessorWithNullParents() {
     }
 
+    @Test(expected = TopologyException.class)
+    public void testAddProcessorWithBadSupplier() {
```
We should also test `addGlobalStateStore()`

soarez replied:
Assuming you mean `.addGlobalStore()`, as there's no `.addGlobalStateStore()` in the builder. There was a supplier usage check missing there too.

mjsax replied:
Yes. Sorry for using the wrong name...

mjsax reviewed (commented):
Thanks for the PR @soarez -- I closed the other PR (obviously you cannot close it as you don't own it...).

Did a pass. Sorry for the delay. We have a high review load and it sometimes takes a while. Feel free to ping regularly to remind me on reviewing...

soarez commented:
Thanks for the review @mjsax. I think I've addressed your feedback and I have a question about whether or not we should extend this to `ValueTransformerSupplier`.

soarez commented:
Pinging @mjsax. Don't forget to review this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamUtil.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * @throws IllegalArgumentException if the same transformer instance is obtained each time
+     */
+    static void checkSupplier(final TransformerSupplier<?, ?, ?> supplier) {
```
I am wondering if we actually need three methods? Could we use `java.util.function.Supplier` instead (we don't really care about generic types.

To customize the error message we just pass an additional `String` or use `supplier.getClass().getName()` ?

soarez replied:
Good idea. `supplier.getClass().getName()` doesn't really work, but we get the name by looking through `supplier.getClass().getInterfaces()`.

soarez committed:
KAFKA-10036 Detect singleton suppliers

soarez commented:
Squashed and rebased as there were conflicts. Please take another look @mjsax .

mjsax commented:
@soarez Sorry for dropping the ball on this PR. Seems there are new conflicts. Can you rebase again?

soarez commented:
@mjsax rebased now. Can you give this another look?

soarez commented:
Seems I had forgotten to run `:streams:test-utils:unitTest`. Tests should be passing now. Please take another look @mjsax

mjsax commented:
Retest this please.

mjsax commented:
@soarez Sorry but you will need to rebase your PR to get https://github.com/apache/kafka/commit/2db67db8e1329cb2e047322cff81d97ff98b4328 -- otherwise, Jenkins does fail...

soarez commented:
@mjsax rebased and fixed an error. Can we try running the tests again?

mjsax commented:
The build did run, but failed with a compile error: Maybe something wrong with the rebase you did?
```
/home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-9000@2/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableTransformValuesTest.java:209: error: cannot find symbol
          new KTableTransformValues<>(parent, new SingletonNoOpValueTransformer<>(), null).enableSendingOldValues(true);
```

soarez committed:
Missed compilation error

soarez commented:
Yes, sorry for that @mjsax . I thought I had already pushed the fix. Not enough ☕ 

Fixed it but seems one test still failed for JDK 8 - `org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta` - but it seems unrelated to this patch and [known to be flaky](https://github.com/apache/kafka/pull/9466#issuecomment-713246622).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java:
```diff
@@ -33,12 +38,17 @@
  * @see TransformerSupplier
  * @see KStream#transform(TransformerSupplier, String...)
  */
-public interface ValueTransformerSupplier<V, VR> extends ConnectedStoreProvider {
+public interface ValueTransformerSupplier<V, VR> extends ConnectedStoreProvider, Supplier<ValueTransformer<V, VR>> {
```
This seems to be a public API change that we cannot do without a KIP. Seem you added it so you can pass the different suppliers into `checkSupplier` ? Also not sure if `checkSupplier` must be as "complicated" as proposed.

soarez replied:
Indeed, that was why. Since it already conforms to the interface, I didn't realize this would be a public API change. Will revert and overload `checkSupplier` instead.

mjsax replied:
Yeah. Even if it's "compatible" and does not break anything, it's still fall into the "public api change" category...

mjsax commented on streams/src/main/java/org/apache/kafka/streams/internals/ApiUtils.java:
```diff
@@ -75,4 +81,31 @@ public static long validateMillisecondInstant(final Instant instant, final Strin
+     */
+    public static void checkSupplier(final Supplier<?> supplier) {
+        if (supplier.get() == supplier.get()) {
+            final String supplierClass = getAllImplementedInterfaces(supplier.getClass()).stream()
```
Do we really need to try to extract the concrete interface name?

soarez replied:
We don't _really_ need to. I thought it could make for a better, and also predictable, error message. But maybe just using the implementing class name is fine.

mjsax replied:
> But maybe just using the implementing class name is fine.

That was my though, too.

soarez committed:
Overload and simplify checkSupplier

soarez commented:
Thanks for giving this another look @mjsax

mjsax commented:
Thanks for the PR @soarez! Merged to `trunk`.

Took way to long to get this merged. Sorry for the delay.

soarez commented:
> Took way to long to get this merged.

My fault for screwing up several times. Thanks for not giving up on this and all the time dedicated to reviewing.
